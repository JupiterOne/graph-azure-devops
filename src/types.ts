import { IntegrationInstanceConfig } from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface ADOIntegrationConfig extends IntegrationInstanceConfig {
  /**
   * https://dev.azure.com/ + an Azure Boards organization name (ex: "https://dev.azure.com/jupiterone")
   */
  orgUrl: string;

  /**
   * The personal access token used to authenticate requests.
   * Can be generated at {{`orgUrl`}}/_usersSettings/tokens.
   * Requires `Read` permissions for `Work Items` and `Project and Team`
   */
  accessToken: string;
  alertSeverities?: string;
}
