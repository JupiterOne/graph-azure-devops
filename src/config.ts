import { IntegrationIngestionConfigFieldMap } from '@jupiterone/integration-sdk-core';
import { INGESTION_SOURCE_IDS } from './constants';

export const ingestionConfig: IntegrationIngestionConfigFieldMap = {
  [INGESTION_SOURCE_IDS.BUILD]: {
    title: 'Build settings',
    description: 'Gather information related to the projects build settings.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.ALERTS]: {
    title: 'Devops alerts',
    description: 'Gather information about repository alerts.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.SERVICE]: {
    title: 'Devops service',
    description:
      'Creates structural relationships between repositories and projects.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.ENVIRONMENTS]: {
    title: 'Environments',
    description: 'Ingest devops project environments.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.PIPELINES]: {
    title: 'Pipelines',
    description: 'Ingest devops project pipelines.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.PROJECTS]: {
    title: 'Projects',
    description: 'Ingest devops projects.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.PULL_REQUESTS]: {
    title: 'Pull Requests',
    description: 'Ingest pull requests and information about their state.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.REPOS]: {
    title: 'Repositories',
    description: 'Ingest azure devops repositories.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.TEAMS]: {
    title: 'Teams',
    description: 'Ingest teams and links users that belong to them.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.USERS]: {
    title: 'Users',
    description: 'Ingest azure devops users.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.WORK_ITEMS]: {
    title: 'Work items',
    description: 'Ingest work items.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
};
