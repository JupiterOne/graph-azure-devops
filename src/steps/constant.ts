import {
  RelationshipClass,
  RelationshipDirection,
  StepEntityMetadata,
  StepMappedRelationshipMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const PullRequestStatus = {
  /**
   * Status not set. Default state.
   */
  0: 'NotSet',
  /**
   * Pull request is active.
   */
  1: 'Active',
  /**
   * Pull request is abandoned.
   */
  2: 'Abandoned',
  /**
   * Pull request is completed.
   */
  3: 'Completed',
  /**
   * Used in pull request search criteria to include all statuses.
   */
  4: 'All',
};

export const AlertType = {
  /**
   * The code has an unspecified vulnerability type
   */
  0: 'Unknown',
  /**
   * The code uses a dependency with a known vulnerability.
   */
  1: 'Dependency',
  /**
   * The code contains a secret that has now been compromised and must be revoked.
   */
  2: 'Secret',
  /**
   * The code contains a weakness determined by static analysis.
   */
  3: 'Code',
};

export const AlertState = {
  /**
   * Alert is in an indeterminate state
   */
  0: 'Unknown',
  /**
   * Alert has been detected in the code
   */
  1: 'Active',
  /**
   * Alert was dismissed by a user
   */
  2: 'Dismissed',
  /**
   * The issue is no longer detected in the code
   */
  4: 'Fixed',
  /**
   * The tool has determined that the issue is no longer a risk
   */
  8: 'AutoDismissed',
};

export const AlertSeverity = {
  0: 'Low',
  1: 'Medium',
  2: 'High',
  3: 'Critical',
  4: 'Note',
  5: 'Warning',
  6: 'Error',
};

export const Steps = {
  FETCH_SERVICE: 'fetch-devops-service',
  BUILD_SERVICE_ACCOUNT_RELATIONSHIP:
    'build-devOps-account-service-relationship',
  FETCH_PROJECTS: 'fetch-projects',
  FETCH_ACCOUNT: 'fetch-accounts',
  FETCH_BUILD_SETTINGS: 'fetch-build-settings',
  BUILD_PROJECT_BUILD_SETTING_RELATIONSHIP:
    'build-project-build-setting-relationship',
  FETCH_PIPELINE: 'fetch-pipelines',
  BUILD_PROJECT_PIPELINE_RELATIONSHIP: 'build-project-pipeline-relationship',
  FETCH_PULL_REQUEST: 'fetch-pullrequests',
  FETCH_REPOSITORY: 'fetch-repositories',
  BUILD_REPO_PR_RELATIONSHIP: 'build-repo-pr-relationship',
  BUILD_ACCONT_REPO_RELATIONSHIP: 'build-account-repo-relationship',
  BUILD_USER_OPENED_PR_RELATIONSHIP: 'build-user-opened-pr-relationship',
  FETCH_USERS: 'fetch-users',
  BUILD_USER_REVIEWED_PR_RELATIONSHIP: 'build-user-reviewd-pr-relationship',
  BUILD_USER_APPROVED_PR_RELATIONSHIP: 'build-user-approved-pr-relationship',
  FETCH_ALERTS: 'fetch-alerts',
  BUILD_REPO_ALERT_RELATIONSHIP: 'build-repo-alert-relationship',
};

export const Entities: Record<
  | 'DEVOPS_SERVICE_ENTITY'
  | 'PROJECT_ENTITY'
  | 'ACCOUNT_ENTITY'
  | 'BUILD_SETTING_ENTITY'
  | 'PIPELINE_ENTITY'
  | 'PULL_REQUEST_ENTITY'
  | 'REPOSITORY_ENTITY'
  | 'USER_ENTITY'
  | 'ALERT_ENTITY',
  StepEntityMetadata
> = {
  DEVOPS_SERVICE_ENTITY: {
    resourceName: 'AzureDevOps',
    _type: 'azure_devops',
    _class: ['Service'],
  },
  PROJECT_ENTITY: {
    resourceName: 'ADO Project',
    _type: 'azure_devops_project',
    _class: 'Project',
  },
  ACCOUNT_ENTITY: {
    resourceName: 'Azure Devops Account',
    _type: 'azure_devops_account',
    _class: 'Account',
  },
  BUILD_SETTING_ENTITY: {
    resourceName: 'AzureBuildSettings',
    _type: 'azure_devops_build_settings',
    _class: 'Configuration',
  },
  PIPELINE_ENTITY: {
    resourceName: 'AzureDevOpsPipeline',
    _type: 'azure_devops_pipeline',
    _class: 'Workflow',
  },
  PULL_REQUEST_ENTITY: {
    resourceName: 'AzureDevOpsPullRequest',
    _type: 'azure_devops_pr',
    _class: 'PR',
  },
  REPOSITORY_ENTITY: {
    resourceName: 'Repository',
    _type: 'azure_devops_repo',
    _class: 'Repository',
  },
  USER_ENTITY: {
    resourceName: 'ADO User',
    _type: 'azure_devops_user',
    _class: 'User',
  },
  ALERT_ENTITY: {
    resourceName: 'Azure DevOps Alerts',
    _type: 'azure_devops_alert_finding',
    _class: 'Finding',
  },
};

export const Relationships: Record<
  | 'ACCOUNT_HAS_USERS'
  | 'USERS_MANAGES_ACCOUNT'
  | 'AZURE_DEVOPS_ACCOUNT_HAS_DEVOPS_SERVICE'
  | 'AZURE_DEVOPS_SCANS_PROJECTS'
  | 'AZURE_DEVOPS_ACCOUNT_HAS_PROJECTS'
  | 'PROJECT_HAS_BUILD_SETTINGS'
  | 'AZURE_DEVOPS_PROJECT_HAS_PIPELINE'
  | 'REPO_HAS_PULL_REQUEST'
  | 'PROJECT_HAS_REPO'
  | 'ACCOUNT_OWNS_REPO'
  | 'USER_OPENED_PR'
  | 'USER_REVIEWED_PR'
  | 'USER_APPROVED_PR'
  | 'REPO_HAS_ALERTS',
  StepRelationshipMetadata
> = {
  ACCOUNT_HAS_USERS: {
    _type: 'azure_devops_account_has_user',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT_ENTITY._type,
    targetType: Entities.USER_ENTITY._type,
  },
  USERS_MANAGES_ACCOUNT: {
    _type: 'azure_devops_user_manages_account',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.USER_ENTITY._type,
    targetType: Entities.ACCOUNT_ENTITY._type,
  },
  AZURE_DEVOPS_ACCOUNT_HAS_DEVOPS_SERVICE: {
    _type: 'azure_devops_account_has_devops',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT_ENTITY._type,
    targetType: Entities.DEVOPS_SERVICE_ENTITY._type,
  },
  AZURE_DEVOPS_SCANS_PROJECTS: {
    _type: 'azure_devops_scans_project',
    _class: RelationshipClass.SCANS,
    sourceType: Entities.DEVOPS_SERVICE_ENTITY._type,
    targetType: Entities.PROJECT_ENTITY._type,
  },
  AZURE_DEVOPS_ACCOUNT_HAS_PROJECTS: {
    _type: 'azure_devops_account_has_project',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT_ENTITY._type,
    targetType: Entities.PROJECT_ENTITY._type,
  },
  PROJECT_HAS_BUILD_SETTINGS: {
    _type: 'azure_devops_project_has_build_settings',
    _class: RelationshipClass.HAS,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: Entities.BUILD_SETTING_ENTITY._type,
  },
  AZURE_DEVOPS_PROJECT_HAS_PIPELINE: {
    _type: 'azure_devops_project_has_pipeline',
    _class: RelationshipClass.HAS,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: Entities.PIPELINE_ENTITY._type,
  },
  REPO_HAS_PULL_REQUEST: {
    _type: 'azure_devops_repo_has_pr',
    _class: RelationshipClass.HAS,
    sourceType: Entities.REPOSITORY_ENTITY._type,
    targetType: Entities.PULL_REQUEST_ENTITY._type,
  },
  PROJECT_HAS_REPO: {
    _type: 'azure_devops_project_uses_repo',
    _class: RelationshipClass.USES,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: Entities.REPOSITORY_ENTITY._type,
  },
  ACCOUNT_OWNS_REPO: {
    _type: 'azure_devops_account_owns_repo',
    _class: RelationshipClass.OWNS,
    sourceType: Entities.ACCOUNT_ENTITY._type,
    targetType: Entities.REPOSITORY_ENTITY._type,
  },
  USER_OPENED_PR: {
    _type: 'azure_devops_user_opened_pr',
    _class: RelationshipClass.OPENED,
    sourceType: Entities.USER_ENTITY._type,
    targetType: Entities.PULL_REQUEST_ENTITY._type,
  },
  USER_REVIEWED_PR: {
    _type: 'azure_devops_user_reviewed_pr',
    _class: RelationshipClass.REVIEWED,
    sourceType: Entities.USER_ENTITY._type,
    targetType: Entities.PULL_REQUEST_ENTITY._type,
  },
  USER_APPROVED_PR: {
    _type: 'azure_devops_user_approved_pr',
    _class: RelationshipClass.APPROVED,
    sourceType: Entities.USER_ENTITY._type,
    targetType: Entities.PULL_REQUEST_ENTITY._type,
  },
  REPO_HAS_ALERTS: {
    _type: 'azure_devops_repo_has_alert_finding',
    _class: RelationshipClass.HAS,
    sourceType: Entities.REPOSITORY_ENTITY._type,
    targetType: Entities.ALERT_ENTITY._type,
  },
};

export const MappedRelationships: Record<
  'PROJECT_USES_GITHUB_REPO' | 'PROJECT_USES_BITBUCKET_REPO',
  StepMappedRelationshipMetadata
> = {
  PROJECT_USES_GITHUB_REPO: {
    _type: 'azure_devops_project_uses_repo',
    _class: RelationshipClass.USES,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: 'github_repo',
    direction: RelationshipDirection.FORWARD,
  },
  PROJECT_USES_BITBUCKET_REPO: {
    _type: 'azure_devops_project_uses_repo',
    _class: RelationshipClass.USES,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: 'bitbucket_repo',
    direction: RelationshipDirection.FORWARD,
  },
};
