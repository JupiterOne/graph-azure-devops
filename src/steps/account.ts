import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { ADOIntegrationConfig } from '../types';
import { Entities, Steps } from './constant';

export function getAccountKey(uniqueId) {
  return `azure-devops-account:${uniqueId}`;
}

export async function fetchAccountDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<ADOIntegrationConfig>) {
  const name = `Azure Devops - ${instance.name}`;
  const accountEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {
          id: 'azure-devops',
          name: 'Azure Devops Account',
        },
        assign: {
          _key: getAccountKey(instance.id),
          _type: Entities.ACCOUNT_ENTITY._type,
          _class: 'Account',
          name,
          displayName: name,
        },
      },
    }),
  );
  await jobState.setData(Entities.ACCOUNT_ENTITY._type, accountEntity);
}

export const accountSteps: IntegrationStep<ADOIntegrationConfig>[] = [
  {
    id: Steps.FETCH_ACCOUNT,
    name: 'Fetch Account Details',
    entities: [Entities.ACCOUNT_ENTITY],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
