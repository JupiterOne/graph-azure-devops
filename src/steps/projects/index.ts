import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  //IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../../client';
import { ADOIntegrationConfig } from '../../types';
import { Entities, Relationships, Steps } from '../constant';
import { INGESTION_SOURCE_IDS } from '../../constants';

export async function fetchProjects({
  instance,
  jobState,
}: IntegrationStepExecutionContext<ADOIntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(
    Entities.ACCOUNT_ENTITY._type,
  )) as Entity;

  await apiClient.iterateProjects(async (project) => {
    const projectEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: project,
          assign: {
            _type: Entities.PROJECT_ENTITY._type,
            _class: Entities.PROJECT_ENTITY._class,
            _key: project.id as string,
            name: project.name,
            displayName: project.name,
            abbreviation: project.abbreviation,
            description: project.description,
            webLink: project.url,
            state: project.state,
            revision: project.revision,
            visibility: project.visibility,
            createdOn: undefined,
            updatedOn: undefined,
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: projectEntity,
      }),
    );
  });
}

export const projectSteps: IntegrationStep<ADOIntegrationConfig>[] = [
  {
    id: Steps.FETCH_PROJECTS,
    name: 'Fetch Projects',
    entities: [Entities.PROJECT_ENTITY],
    relationships: [Relationships.AZURE_DEVOPS_ACCOUNT_HAS_PROJECTS],
    dependsOn: [Steps.FETCH_ACCOUNT],
    executionHandler: fetchProjects,
    ingestionSourceId: INGESTION_SOURCE_IDS.PROJECTS,
  },
];
