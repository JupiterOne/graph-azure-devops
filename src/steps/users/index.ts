import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../../client';
import { ADOIntegrationConfig } from '../../types';
import { Entities, Relationships, Steps } from '../constant';

export const UNIQUE_NAME_TO_USER_ID_MAPPING_PREFIX = 'UserUniqueName:';

export function getUserKey(uniqueId) {
  return uniqueId;
}

export async function fetchUsers({
  logger,
  instance,
  jobState,
}: IntegrationStepExecutionContext<ADOIntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(
    Entities.ACCOUNT_ENTITY._type,
  )) as Entity;

  await apiClient.iterateUsers(async (user) => {
    if (!user.identity?.id) {
      logger.info('Skipping user resource where user.identity is undefined');
      return;
    }
    const userEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: user,
          assign: {
            _type: Entities.USER_ENTITY._type,
            _class: Entities.USER_ENTITY._class,
            _key: getUserKey(user.identity?.id),
            name: user.identity?.displayName,
            displayName: user.identity?.displayName,
            email: user.identity?.uniqueName,
            emailDomain: [user.identity?.uniqueName?.split('@').pop() || ''],
            username: user.identity?.uniqueName,
            webLink: user.identity?.url,
            imageLink: user.identity?.imageUrl,
            descriptor: user.identity?.descriptor,
            profileLink: user.identity?.profileUrl,
            active: true,
          },
        },
      }),
    );

    // Need to be able to look up the userId based on the user's email due to the way that work items are linked to users.
    const key =
      UNIQUE_NAME_TO_USER_ID_MAPPING_PREFIX + user.identity?.uniqueName;
    await jobState.setData(key, user.identity?.id);

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: userEntity,
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.MANAGES,
        from: userEntity,
        to: accountEntity,
      }),
    );
  });
}

export const userSteps: IntegrationStep<ADOIntegrationConfig>[] = [
  {
    id: Steps.FETCH_USERS,
    name: 'Fetch Users',
    entities: [Entities.USER_ENTITY],
    relationships: [
      Relationships.ACCOUNT_HAS_USERS,
      Relationships.USERS_MANAGES_ACCOUNT,
    ],
    dependsOn: [Steps.FETCH_ACCOUNT],
    executionHandler: fetchUsers,
  },
];
