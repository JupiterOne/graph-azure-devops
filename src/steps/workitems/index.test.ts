import { RelationshipClass } from '@jupiterone/data-model';
import {
  createMockStepExecutionContext,
  MockIntegrationStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { fetchWorkItems } from '.';
import { config } from '../../../test/config';
import { setupAzureRecording } from '../../../test/recording';
import { getMatchRequestsBy } from '../../../test/getMatchRequestsBy';
import { testEntities } from '../../../test/testEntities';
import { testRelationships } from '../../../test/testRelationships';
import { toArray } from '../../../test/toArray';
import { ADOIntegrationConfig } from '../../types';
import { fetchAccountDetails } from '../account';
import { fetchProjects } from '../projects';
import { fetchUsers, UNIQUE_NAME_TO_USER_ID_MAPPING_PREFIX } from '../users';
import { Entities, unit_test_custom_timeout } from '../constant';

let recording: Recording;

afterEach(async () => {
  if (recording) {
    await recording.stop();
  }
});

async function setupFetchWorkItemsContext(): Promise<
  MockIntegrationStepExecutionContext<ADOIntegrationConfig>
> {
  const setupContext = createMockStepExecutionContext({
    instanceConfig: config,
  });

  await fetchAccountDetails(setupContext);
  await fetchUsers(setupContext);
  await fetchProjects(setupContext);

  const accountEntity = setupContext.jobState.collectedEntities.filter((e) =>
    toArray(e._class).includes('Account'),
  )[0];
  const userEntity = setupContext.jobState.collectedEntities.filter((e) =>
    toArray(e._class).includes('User'),
  )[0];

  return createMockStepExecutionContext({
    instanceConfig: config,
    entities: setupContext.jobState.collectedEntities,
    relationships: setupContext.jobState.collectedRelationships,
    setData: {
      [Entities.ACCOUNT_ENTITY._type]: accountEntity,
      [UNIQUE_NAME_TO_USER_ID_MAPPING_PREFIX + userEntity.email]:
        userEntity._key,
    },
  });
}

describe('fetchWorkItems', () => {
  it(
    'Should create user entities and relationships',
    async () => {
      recording = setupAzureRecording({
        directory: __dirname,
        name: 'fetchWorkItems',
        options: {
          matchRequestsBy: getMatchRequestsBy({ config: config }),
        },
      });

      const context = await setupFetchWorkItemsContext();
      await fetchWorkItems(context);

      const workItemEntities = context.jobState.collectedEntities;
      testEntities(workItemEntities, 'workItemEntities');
      const createdRelationships =
        context.jobState.collectedRelationships.filter(
          (r) => r._class === RelationshipClass.CREATED,
        );
      testRelationships(createdRelationships, 'createdRelationships');
      const assignedRelationships =
        context.jobState.collectedRelationships.filter(
          (r) => r._class === RelationshipClass.ASSIGNED,
        );
      testRelationships(assignedRelationships, 'assignedRelationships');
      const hasRelationships = context.jobState.collectedRelationships.filter(
        (r) => r._class === RelationshipClass.HAS,
      );
      testRelationships(hasRelationships, 'hasRelationships');
    },
    unit_test_custom_timeout,
  );
});
