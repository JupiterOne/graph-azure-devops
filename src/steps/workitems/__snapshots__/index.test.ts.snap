// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchWorkItems Should create user entities and relationships: assignedRelationships 1`] = `
[
  {
    "_class": "ASSIGNED",
    "_fromEntityKey": "a487aae2-d50b-6c04-90e1-e20dcc3dc9cc",
    "_key": "a487aae2-d50b-6c04-90e1-e20dcc3dc9cc|assigned|af58e917-ee45-445b-bfb7-7850a53fcf211",
    "_toEntityKey": "af58e917-ee45-445b-bfb7-7850a53fcf211",
    "_type": "azure_devops_user_assigned_work_item",
    "displayName": "ASSIGNED",
  },
]
`;

exports[`fetchWorkItems Should create user entities and relationships: createdRelationships 1`] = `
[
  {
    "_class": "CREATED",
    "_fromEntityKey": "a487aae2-d50b-6c04-90e1-e20dcc3dc9cc",
    "_key": "a487aae2-d50b-6c04-90e1-e20dcc3dc9cc|created|af58e917-ee45-445b-bfb7-7850a53fcf211",
    "_toEntityKey": "af58e917-ee45-445b-bfb7-7850a53fcf211",
    "_type": "azure_devops_user_created_work_item",
    "displayName": "CREATED",
  },
]
`;

exports[`fetchWorkItems Should create user entities and relationships: hasRelationships 1`] = `
[
  {
    "_class": "HAS",
    "_fromEntityKey": "af58e917-ee45-445b-bfb7-7850a53fcf21",
    "_key": "af58e917-ee45-445b-bfb7-7850a53fcf21|has|af58e917-ee45-445b-bfb7-7850a53fcf211",
    "_toEntityKey": "af58e917-ee45-445b-bfb7-7850a53fcf211",
    "_type": "azure_devops_project_has_work_item",
    "displayName": "HAS",
  },
]
`;

exports[`fetchWorkItems Should create user entities and relationships: workItemEntities 1`] = `
[
  {
    "_class": [
      "Record",
    ],
    "_key": "af58e917-ee45-445b-bfb7-7850a53fcf211",
    "_rawData": [
      {
        "name": "default",
        "rawData": {
          "fields": {
            "Microsoft.VSTS.Common.Priority": 2,
            "Microsoft.VSTS.Common.StateChangeDate": "2024-03-14T14:35:53.323Z",
            "System.AreaId": 8,
            "System.AreaLevel1": "Project2",
            "System.AreaPath": "Project2",
            "System.AssignedTo": "metron <metron@j1AzureIntegrationDev.onmicrosoft.com>",
            "System.AttachedFileCount": 0,
            "System.AuthorizedAs": "metron <metron@j1AzureIntegrationDev.onmicrosoft.com>",
            "System.AuthorizedDate": "2024-03-14T15:20:53.2Z",
            "System.ChangedBy": "metron <metron@j1AzureIntegrationDev.onmicrosoft.com>",
            "System.ChangedDate": "2024-03-14T15:20:53.2Z",
            "System.CommentCount": 0,
            "System.CreatedBy": "metron <metron@j1AzureIntegrationDev.onmicrosoft.com>",
            "System.CreatedDate": "2024-03-14T14:35:53.323Z",
            "System.ExternalLinkCount": 0,
            "System.HyperLinkCount": 0,
            "System.Id": 1,
            "System.IterationId": 5,
            "System.IterationLevel1": "Project2",
            "System.IterationPath": "Project2",
            "System.NodeName": "Project2",
            "System.PersonId": 246753397,
            "System.Reason": "Added to backlog",
            "System.RelatedLinkCount": 0,
            "System.RemoteLinkCount": 0,
            "System.Rev": 2,
            "System.RevisedDate": "9999-01-01T00:00:00Z",
            "System.State": "To Do",
            "System.TeamProject": "Project2",
            "System.Title": "test workitem",
            "System.Watermark": 2,
            "System.WorkItemType": "Epic",
          },
          "id": 1,
          "projectId": "af58e917-ee45-445b-bfb7-7850a53fcf21",
          "rev": 2,
        },
      },
    ],
    "_type": "azure_devops_work_item",
    "areaPath": "Project2",
    "changedBy": "metron <metron@j1AzureIntegrationDev.onmicrosoft.com>",
    "changedDate": "2024-03-14T15:20:53.2Z",
    "commentCount": 0,
    "createdDate": "2024-03-14T14:35:53.323Z",
    "createdOn": undefined,
    "description": undefined,
    "displayName": "test workitem",
    "history": undefined,
    "interationPath": "Project2",
    "name": "test workitem",
    "priority": 2,
    "projectId": "af58e917-ee45-445b-bfb7-7850a53fcf21",
    "projectName": "Project2",
    "reason": "Added to backlog",
    "revision": 2,
    "state": "To Do",
    "stateChangeDate": "2024-03-14T14:35:53.323Z",
    "teamProject": "Project2",
    "type": "Epic",
    "webLink": undefined,
    "workItemType": "Epic",
  },
]
`;
